File: directory_analyzer/csharp-ver/dira/types/info.cs
using System.Diagnostics.CodeAnalysis;
using System.Globalization;

namespace dira.types
{

    public class AnalyzedInfo
    {
        public AnalyzedInfo()
        {
        }

        /// <summary>
        /// Count of directories found during the analysis
        /// </summary>
        public required uint FoundDirs { get; set; }

        /// <summary>
        /// Count of files found during the analysis
        /// </summary>
        public required uint FoundFiles { get; set; }

        /// <summary>
        /// Info about symlinks found during the analysis
        /// </summary>
        public SymlinkInfo? FoundSymlinks { get; set; }

        /// <summary>
        /// Info about files grouped by file type
        /// </summary>
        public Dictionary<FileExtension, FileTypeInfo>? FileInfo { get; set; }

        /// <summary>
        /// How many bytes taken up by all the files
        /// </summary>
        public required ulong TotalBytes { get; set; }


        /// <summary>
        /// Calculates the the percent of total files and percent of total size on disk for each type of file,if file info is gathered.
        /// </summary>
        public void CalculatePercentagesForInfo()
        {
            if (FileInfo != null)
            {
                foreach (var info in FileInfo)
                {
                    info.Value.CalculatePercentages(TotalBytes, FoundFiles);
                }
            }
        }

        public override bool Equals(object? obj)
        {
            if (obj == null || obj is not AnalyzedInfo)
            {
                return false;
            }
            var other = (AnalyzedInfo)obj;
            bool dicts = true;
            if (FileInfo != null && other.FileInfo != null)
            {
                dicts = other.FileInfo.All(kv =>
                {
                    if (FileInfo.TryGetValue(kv.Key, out FileTypeInfo? value))
                    {
                        return value == kv.Value;
                    }
                    else
                    {
                        return false;
                    }
                });
            }
            bool syms = this.FoundSymlinks == other.FoundSymlinks;

            return this.FoundDirs == other.FoundDirs && this.FoundFiles == other.FoundFiles && this.TotalBytes == other.TotalBytes && syms && dicts;
        }

        public static bool operator ==(AnalyzedInfo left, AnalyzedInfo right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(AnalyzedInfo left, AnalyzedInfo right)
        {
            return !(left == right);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(FoundDirs, FoundFiles, FoundSymlinks, FileInfo, TotalBytes);
        }

        public override string ToString()
        {
            string symlinks_str = "";
            if (FoundSymlinks != null)
            {
                symlinks_str = String.Format("\n{0}", FoundSymlinks.ToString());
            }

            string info_str = "";
            if (FileInfo != null)
            {
                foreach ((var file_ext, var ft_info) in FileInfo)
                {
                    info_str += String.Format("\nFile extension:{0}{1}", file_ext.ToString(), ft_info.ToString());
                }
            }
            var str = String.Format("Found directories:{0}\nFound files:{1}\nTotal bytes:{2} bytes{3}\n{4}", FoundDirs, FoundFiles, TotalBytes, symlinks_str, info_str);
            return str;
        }
    }

    public readonly struct FileExtension(string ext)
    {
        public string Ext { get; } = ext;

        public override string ToString()
        {
            return Ext.ToString();
        }
    }

    public class FileTypeInfo(uint num_files, FileTypeInfoRecords largest_file, FileTypeInfoRecords smallest_file, ulong size_in_bytes, float? percent_of_total_files, float? percent_of_total_size)
    {
        /// <summary>
        /// Number of files of this type
        /// </summary>
        public uint NumFiles { get; set; } = num_files;

        public FileTypeInfoRecords LargestFile { get; set; } = largest_file;

        public FileTypeInfoRecords SmallestFile { get; set; } = smallest_file;

        /// <summary>
        /// Total size of all files of this type
        /// </summary>
        public ulong SizeInBytes { get; set; } = size_in_bytes;

        public float? PercentOfTotalFiles { get; private set; } = percent_of_total_files;

        public float? PercentOfTotalSize { get; private set; } = percent_of_total_size;

        /// <summary>
        /// Calculates PercentOfTotalFiles and PercentOfTotalSize
        /// </summary>
        /// <param name="total_bytes"></param>
        /// <param name="total_files"></param>
        public void CalculatePercentages(ulong total_bytes, uint total_files)
        {
            PercentOfTotalFiles = (float)NumFiles / (float)total_files;
            PercentOfTotalSize = (float)SizeInBytes / (float)total_bytes;
        }

        public override bool Equals(object? obj)
        {
            if (obj != null && obj is FileTypeInfo other)
            {
                return this.NumFiles == other.NumFiles && this.LargestFile == other.LargestFile && this.SmallestFile == other.SmallestFile && this.SizeInBytes == other.SizeInBytes && this.PercentOfTotalFiles == other.PercentOfTotalFiles && this.PercentOfTotalSize == other.PercentOfTotalSize;
            }
            return false;
        }

        public static bool operator ==(FileTypeInfo left, FileTypeInfo right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(FileTypeInfo left, FileTypeInfo right)
        {
            return !(left == right);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(NumFiles, LargestFile, SmallestFile, SizeInBytes, PercentOfTotalFiles, PercentOfTotalSize);
        }

        public override string ToString()
        {
            (var per_tot_files, var per_tot_size) = PercentagesInString();
            var per_tot_files_str = string.Format("  % of total files: {0}\n", per_tot_files);
            var per_tot_size_str = string.Format("\n  % of total size: {0}", per_tot_size);


            return string.Format("\n  Number of files:{0}\n{1}  Largest file: {2}\n  Smallest file: {3}\n  Size in bytes for this type: {4}{5}", NumFiles, per_tot_files_str, LargestFile, SmallestFile, SizeInBytes, per_tot_size_str);
        }

        public (PercentOfTotalFiles, PercentOfTotalSize) PercentagesInString()
        {
            var per_tot_files = new PercentOfTotalFiles("N/A");
            if (PercentOfTotalFiles != null)
            {
                per_tot_files = new PercentOfTotalFiles(string.Format("{0:.##}", PercentOfTotalFiles * 100.0));
                if (per_tot_files.Percent == "")
                {
                    per_tot_files = new PercentOfTotalFiles("< 0.01");
                }
            }
            var per_tot_size = new PercentOfTotalSize("N/A");
            if (PercentOfTotalSize != null)
            {
                per_tot_size = new PercentOfTotalSize(string.Format("{0:.##}", PercentOfTotalSize * 100.0));
                if (per_tot_size.Percent == "")
                {
                    per_tot_size = new PercentOfTotalSize("< 0.01");
                }
            }
            return (per_tot_files, per_tot_size);
        }


    }

    public readonly struct PercentOfTotalFiles(string percent)
    {
        public string Percent { get; } = percent;

        public override string ToString()
        {
            return Percent;
        }
    }

    public readonly struct PercentOfTotalSize(string percent)
    {
        public string Percent { get; } = percent;

        public override string ToString()
        {
            return Percent;
        }
    }
    /// <summary>
    /// This is for holding info about specific notable files
    /// </summary>
    public readonly struct FileTypeInfoRecords(string path, ulong size)
    {
        /// <summary>
        /// Path to the file
        /// </summary>
        public string Path { get; } = path;

        /// <summary>
        /// The size of the file
        /// </summary>
        public ulong Size { get; } = size;

        public override bool Equals([NotNullWhen(true)] object? obj)
        {
            if (obj != null && obj is FileTypeInfoRecords other)
            {

                return this.Path == other.Path && this.Size == other.Size;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Path, Size);
        }

        public static bool operator ==(FileTypeInfoRecords left, FileTypeInfoRecords right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(FileTypeInfoRecords left, FileTypeInfoRecords right)
        {
            return !(left == right);
        }

        public override string ToString()
        {
            return string.Format("\n    Path:{0}\n    Size: {1}", Path, Size);
        }
    }

    public struct SymlinkInfo
    {

        public uint FoundSymlinks { get; private set; }

        /// <summary>
        /// Number of symlinks that point to files
        /// </summary>
        private uint fileSymlinks;

        /// <summary>
        /// Number of symlinks that point to files
        /// </summary>
        /// <returns></returns>
        public readonly uint GetFileSymlinks()
        {
            return fileSymlinks;
        }

        public void SetFileSymlinks(uint value)
        {
            fileSymlinks = value;
            FoundSymlinks = fileSymlinks + dirSymlinks;
        }

        /// <summary>
        /// Number of symlinks that point to directories
        /// </summary>
        private uint dirSymlinks;

        /// <summary>
        /// Number of symlinks that point to directories
        /// </summary>
        /// <returns></returns>
        public readonly uint GetDirSymlinks()
        {
            return dirSymlinks;
        }

        public void SetDirSymlinks(uint value)
        {
            dirSymlinks = value;
            FoundSymlinks = fileSymlinks + dirSymlinks;
        }

        public override readonly bool Equals([NotNullWhen(true)] object? obj)
        {
            if (obj != null && obj is SymlinkInfo other)
            {
                return this.dirSymlinks == other.dirSymlinks && this.fileSymlinks == other.fileSymlinks && this.FoundSymlinks == other.FoundSymlinks;
            }
            return false;
        }

        public override readonly int GetHashCode()
        {
            return HashCode.Combine(dirSymlinks, fileSymlinks, FoundSymlinks);
        }

        public static bool operator ==(SymlinkInfo left, SymlinkInfo right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(SymlinkInfo left, SymlinkInfo right)
        {
            return !(left == right);
        }

        public override readonly string ToString()
        {
            return string.Format("\nFound symbolic links:{0}\nSymlinks that point to files: {1}\nSymlinks that point to directories: {2}", FoundSymlinks, fileSymlinks, dirSymlinks);
        }
    }
}

File: directory_analyzer/rust-ver/src/types/info.rs
use std::{
    collections::HashMap,
    path::{Path, PathBuf},
};

use serde::Serialize;

use super::FileExtension;

#[derive(Debug, Clone, Default, PartialEq, Serialize)]
pub struct AnalyzedInfo {
    ///Count of directories found during the analysis
    found_dirs: u32,
    ///Count of files found during the analysis
    found_files: u32,
    ///Info about symlinks found during the analysis
    found_symlinks: Option<SymlinkInfo>,
    ///Info about files grouped by file type
    file_info: Option<HashMap<FileExtension, FileTypeInfo>>,
    ///How many bytes taken up by all the files
    total_bytes: u64,
}

impl std::fmt::Display for AnalyzedInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let symlinks_str = match self.found_symlinks {
            Some(sym) => format!("\n{sym}"),
            None => "".to_string(),
        };
        let info_str = match &self.file_info {
            Some(info) => {
                let mut out = String::new();
                for (file_ext, ft_info) in info {
                    out.push_str(&format!("\nFile extension:{file_ext}{ft_info}"));
                }
                out
            }
            None => "".to_string(),
        };
        let str = format!(
            "Found directories: {}\nFound files: {}\nTotal bytes: {} bytes{symlinks_str}\n{info_str}",
            self.found_dirs, self.found_files, self.total_bytes
        );
        write!(f, "{str}")
    }
}

impl AnalyzedInfo {
    pub fn new(
        found_dirs: u32,
        found_files: u32,
        found_symlinks: Option<SymlinkInfo>,
        file_info: Option<HashMap<FileExtension, FileTypeInfo>>,
        total_bytes: u64,
    ) -> Self {
        Self {
            found_dirs,
            found_files,
            found_symlinks,
            file_info,
            total_bytes,
        }
    }

    ///Count of directories found during the analysis
    pub fn found_dirs(&self) -> u32 {
        self.found_dirs
    }
    ///Count of files found during the analysis
    pub fn found_files(&self) -> u32 {
        self.found_files
    }
    ///Info about files grouped by file type
    pub fn file_info(&self) -> Option<&HashMap<FileExtension, FileTypeInfo>> {
        self.file_info.as_ref()
    }

    pub fn found_symlinks(&self) -> Option<&SymlinkInfo> {
        self.found_symlinks.as_ref()
    }

    pub fn found_dirs_mut(&mut self) -> &mut u32 {
        &mut self.found_dirs
    }

    pub fn found_files_mut(&mut self) -> &mut u32 {
        &mut self.found_files
    }

    pub fn set_found_symlinks(&mut self, found_symlinks: Option<SymlinkInfo>) {
        self.found_symlinks = found_symlinks;
    }

    pub fn set_file_info(&mut self, file_info: Option<HashMap<FileExtension, FileTypeInfo>>) {
        self.file_info = file_info;
    }

    ///Returns a mutable reference to found_symlinks if it is Some()
    pub fn found_symlinks_mut(&mut self) -> Option<&mut SymlinkInfo> {
        if let Some(symlink) = &mut self.found_symlinks {
            return Some(symlink);
        }
        None
    }

    ///Returns a mutable reference to file_info if it is Some()
    pub fn file_info_mut(&mut self) -> Option<&mut HashMap<FileExtension, FileTypeInfo>> {
        if let Some(file_info) = &mut self.file_info {
            return Some(file_info);
        }
        None
    }

    ///Calculates the the percent of total files and percent of total size on disk for each type of file,if file info is gathered.
    pub fn calculate_percentages_for_info(&mut self) {
        if let Some(map) = &mut self.file_info {
            for info in map.values_mut() {
                info.calculate_percentages(self.total_bytes, self.found_files);
            }
        }
    }

    ///How many bytes taken up by all the files
    pub fn total_bytes(&self) -> u64 {
        self.total_bytes
    }

    pub fn total_bytes_mut(&mut self) -> &mut u64 {
        &mut self.total_bytes
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Default)]
pub struct FileTypeInfo {
    num_files: u32,
    largest_file: FileTypeInfoRecords,
    smallest_file: FileTypeInfoRecords,
    ///Total size of all files of this type
    size_in_bytes: u64,
    percent_of_total_files: Option<f32>,
    percent_of_total_size: Option<f32>,
}

impl std::fmt::Display for FileTypeInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let (per_tot_files, per_tot_size) = self.percentages_in_string();
        let per_tot_files = format!(" % of total files: {per_tot_files}\n");
        let per_tot_size = format!("\n  % of total size: {per_tot_size}");
        write!(
            f,
            "\n  Number of files:{}\n{}  Largest file: {}\n  Smallest file: {}\n  Size in bytes for this type: {}{}",
            self.num_files,per_tot_files, self.largest_file, self.smallest_file, self.size_in_bytes,per_tot_size
        )
    }
}

///This is for holding info about specific notable files
#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize)]
pub struct FileTypeInfoRecords {
    ///Path to the path
    path: PathBuf,
    ///The size of the file
    size: u64,
}

impl std::fmt::Display for FileTypeInfoRecords {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "\n    Path:{}\n    Size: {}",
            self.path.to_string_lossy(),
            self.size
        )
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default, Copy, Serialize)]
pub struct SymlinkInfo {
    found_symlinks: u32,
    ///Number of symlinks that point to files
    file_symlinks: u32,
    ///Number of symlinks that point to directories
    dir_symlinks: u32,
}

impl std::fmt::Display for SymlinkInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f,"\nFound symbolic links:{}\nSymlinks that point to files: {}\nSymlinks that point to directories: {}",self.found_symlinks,self.file_symlinks,self.dir_symlinks)
    }
}

impl SymlinkInfo {
    pub fn new(found_symlinks: u32, file_symlinks: u32, dir_symlinks: u32) -> Self {
        Self {
            found_symlinks,
            file_symlinks,
            dir_symlinks,
        }
    }

    ///Number of symlinks that point to files
    pub fn file_symlinks(&self) -> u32 {
        self.file_symlinks
    }

    pub fn found_symlinks(&self) -> u32 {
        self.found_symlinks
    }

    ///Number of symlinks that point to directories
    pub fn dir_symlinks(&self) -> u32 {
        self.dir_symlinks
    }

    pub fn found_symlinks_mut(&mut self) -> &mut u32 {
        &mut self.found_symlinks
    }

    pub fn file_symlinks_mut(&mut self) -> &mut u32 {
        &mut self.file_symlinks
    }

    pub fn dir_symlinks_mut(&mut self) -> &mut u32 {
        &mut self.dir_symlinks
    }
}

impl FileTypeInfoRecords {
    pub fn new(path: PathBuf, size: u64) -> Self {
        Self { path, size }
    }

    ///Path to the path
    pub fn path(&self) -> &Path {
        self.path.as_path()
    }
    ///The size of the file
    pub fn size(&self) -> u64 {
        self.size
    }
}

impl FileTypeInfo {
    pub fn new(
        size_in_bytes: u64,
        num_files: u32,
        largest_file: FileTypeInfoRecords,
        smallest_file: FileTypeInfoRecords,
    ) -> Self {
        Self {
            num_files,
            largest_file,
            smallest_file,
            size_in_bytes,
            percent_of_total_files: None,
            percent_of_total_size: None,
        }
    }

    pub fn size_in_bytes(&self) -> u64 {
        self.size_in_bytes
    }

    pub fn num_files(&self) -> u32 {
        self.num_files
    }

    pub fn largest_file(&self) -> &FileTypeInfoRecords {
        &self.largest_file
    }

    pub fn smallest_file(&self) -> &FileTypeInfoRecords {
        &self.smallest_file
    }

    pub fn num_files_mut(&mut self) -> &mut u32 {
        &mut self.num_files
    }

    pub fn size_in_bytes_mut(&mut self) -> &mut u64 {
        &mut self.size_in_bytes
    }

    pub fn set_largest_file(&mut self, largest_file: FileTypeInfoRecords) {
        self.largest_file = largest_file;
    }

    pub fn set_smallest_file(&mut self, smallest_file: FileTypeInfoRecords) {
        self.smallest_file = smallest_file;
    }

    ///Calculates `percent_of_total_files` and `percent_of_total_size`
    pub fn calculate_percentages(&mut self, total_bytes: u64, total_files: u32) {
        self.percent_of_total_files = Some(self.num_files as f32 / total_files as f32);
        self.percent_of_total_size = Some(self.size_in_bytes as f32 / total_bytes as f32);
    }

    pub fn percentages_in_string(&self) -> (PercentageOfFiles, PercentageOfSize) {
        let per_tot_files = match self.percent_of_total_files {
            Some(per) => {
                let tmp = format!("{:.2}", per * 100.0);
                if tmp == "0.00" {
                    "< 0.01".to_string()
                } else {
                    tmp
                }
            }
            None => "N/A\n".to_string(),
        };
        let per_tot_size = match self.percent_of_total_size {
            Some(per) => {
                let tmp = format!("{:.2}", per * 100.0);
                if tmp == "0.00" {
                    "< 0.01".to_string()
                } else {
                    tmp
                }
            }
            None => "N/A".to_string(),
        };
        (per_tot_files, per_tot_size)
    }
}

pub type PercentageOfFiles = String;
pub type PercentageOfSize = String;
